### PR TITLE
Huge Refactor + Code Cleanup + Remove Precompile

### PR DESCRIPTION
**Pull request by nico-bellante**
# Major changes in this PR 
## Removal of Precompile
This was done by the introduction of the `<PageSwitcher/>` component. (See [this comment](https://github.com/OfficeDev/script-lab-react/pull/483#discussion_r243401439)). This allows us to have multiple "pages" on the same domain all using the same index.html, therefore being able to utilize CRA's build tooling. This is done via client-side routing using React Router.

## Custom Functions Refactor
After splitting things up into pages, it became clear that the Custom Functions code shouldn't be a part of the same store as the Editor. Because the logic for CF alone isn't that large, I refactored it into a React Component. For more details, see [this comment](https://github.com/OfficeDev/script-lab-react/pull/483#discussion_r243404305).

## Header and Footer Selectors Refactor
See [this comment](https://github.com/OfficeDev/script-lab-react/pull/483#discussion_r243405967)

## Using more common components 
Additionally, I've refactored a lot of the Editor's code to use the common components created when making the Runner.

# Other Notes:
@Zlatkovsky please try building this locally and going over the TESTING.md scenarios before merging. This refactor went almost too perfectly in the sense that I was shocked things didn't break more while moving everything around, so it would be best to get a second set of eyes looking over everything.

### CODE REVIEW

**Comment by Zlatkovsky** [comment]
Mental note:  we should make sure that the deploy scripts (and other things from the last 2.5 weeks....) get into this branch...

**Comment by Zlatkovsky** [comment]
@nico-bellante , I'm seeing the following warnings when I run:
```
Listening on port 5000
Starting the development server...
Starting the development server...

error TS5055: Cannot write file '.../packages/common/lib/constants.d.ts' because it would overwrite input file.

error TS5055: Cannot write file '.../packages/common/lib/environment.d.ts' because it would overwrite input file.

error TS5055: Cannot write file '.../packages/common/lib/utilities/ensure.fresh.local.storage.d.ts' because it would overwrite input file.

error TS5055: Cannot write file '.../packages/common/lib/utilities/script-loader/constants.d.ts' because it would overwrite input file.

[14:21:05] Found 4 errors. Watching for file changes.
Compiled with warnings.
```

Are you seeing that?  Any ideas why?

File: packages/runner/public/index.html [hunk 1]
```diff
@@ -26,12 +26,173 @@
...
+        }
+
+        40% {
+          -webkit-transform: scale(1);
+          transform: scale(1);
+        }
+      }
+    </style>
+
+    <!-- TODO: Move this out to be dynamically loaded for CFRunner -->
```
Comment (Zlatkovsky) [inline packages/runner/public/index.html]: @nico-bellante , I can help you with this.  Let's do it together, probably.

File: packages/editor/src/pages/Editor/store/messageBar/reducer.ts [hunk 2]
```diff
@@ -106,16 +103,6 @@ const messageBarReducer = (
         link: null,
       };
 
-    case getType(customFunctions.registerMetadata.failure):
```
[the remark this answers is not in the document]
Comment (nico-bellante) [reply packages/editor/src/pages/Editor/store/messageBar/reducer.ts]: more of it was a side effect of bringing custom functions outside of redux

File: packages/editor/package.json [hunk 3]
```diff
@@ -19,10 +19,10 @@
     "moment": "^2.22.2",
     "monaco-editor": "^0.14.3",
     "normalizr": "^3.2.4",
-    "office-ui-fabric-react": "6.79.0",
+    "office-ui-fabric-react": "^6.114.0",
     "query-string": "5.1.1",
-    "react": "^16.5.2",
-    "react-dom": "^16.5.2",
+    "react": "^16.7.0-alpha.2",
```
[the remark this answers is not in the document]
Comment (Zlatkovsky) [reply packages/editor/package.json]: Just create a backlog bug for us to get onto a more stable build once it's ready.  And let's check on it in a couple weeks (for that matter, maybe there already is a newer version...)

File: packages/editor/public/functions.html [hunk 4]
```diff
@@ -2,28 +2,10 @@
 <html>
   <head>
     <meta charset="utf-8" />
-    <title>Script Lab - Add-in Commands</title>
+    <title>Script Lab</title>
+    <meta http-equiv="refresh" content="0; URL='./#/functions'" />
```
[the remark this answers is not in the document]
Comment (Zlatkovsky) [reply packages/editor/public/functions.html]: In reality, so long as we have Script Lab 2017 that we want to keep alive and workign with the same manifest, that won't be the case.
I think it's fine as is.

**Pull request merged**